README.md
# WARNING: This repository is no longer maintained :warning:

> The functionality of this generator is now available in our [MEAN](https://github.com/IBM/mern-app) and [MERN](https://github.com/IBM/mern-app) static applications.

> This repository will not be updated. The repository will be kept available in read-only mode.

# IBM Web Generator


[![IBM Cloud powered][img-ibmcloud-powered]][url-cloud]
[![Travis][img-travis-master]][url-travis-master]
[![Coveralls][img-coveralls-master]][url-coveralls-master]
[![Codacy][img-codacy]][url-codacy]
[![Version][img-version]][url-npm]
[![DownloadsMonthly][img-npm-downloads-monthly]][url-npm]
[![DownloadsTotal][img-npm-downloads-total]][url-npm]
[![License][img-license]][url-npm]
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)

[img-ibmcloud-powered]: https://img.shields.io/badge/IBM%20Cloud-powered-blue.svg
[url-cloud]: http://cloud.ibm.com

[url-npm]: https://www.npmjs.com/package/generator-ibm-web
[img-license]: https://img.shields.io/npm/l/generator-ibm-web.svg
[img-version]: https://img.shields.io/npm/v/generator-ibm-web.svg
[img-npm-downloads-monthly]: https://img.shields.io/npm/dm/generator-ibm-web.svg
[img-npm-downloads-total]: https://img.shields.io/npm/dt/generator-ibm-web.svg

[img-travis-master]: https://travis-ci.org/ibm-developer/generator-ibm-web.svg?branch=master
[url-travis-master]: https://travis-ci.org/ibm-developer/generator-ibm-web/branches

[img-coveralls-master]: https://coveralls.io/repos/github/ibm-developer/generator-ibm-web/badge.svg
[url-coveralls-master]: https://coveralls.io/github/ibm-developer/generator-ibm-web


[img-codacy]: https://api.codacy.com/project/badge/Grade/de3305c225024fe5b551d9f8fee3b587?branch=master
[url-codacy]: https://www.codacy.com/app/ibm-developer/generator-ibm-web



## Getting Started with a generated project

1. Run the generator and choose the options:
  `yo ibm-web`

1. Select a web framework: None, Angular or React

1. Select a language: Swift, Node, Java, Go, Python-Flask or Python-Django

## Frameworks

Frameworks are chosen using the `framework` option at the end of the generator. For example,

```
yo ibm-web --framework {None, React, AngularJS}
```

### Basic

In the basic web project, the following files are created and stored in the `public` directory:

- `public/index.html` - Basic landing page with inline CSS and encoded image assets
- `public/404.html` - Basic 404 error page
- `public/500.html` - Basic 500 error page

It is important that the web server is registered to serve files in the `public` directory.

## React

- [react](https://facebook.github.io/react/) - to build user interfaces
- [webpack](https://webpack.github.io/) - bundles your JS modules
  - sass-loader
  - css-loader
  - style-loader

Web sources for React projects are stored in the `client` directory:

- `client/index.html` - Landing page source
- `client/index.jsx` - Landing page source
- `client/app/App.jsx` - Component injected into landing page body
- `client/404.html` and `client/500.html` - Error page sources
- `client/default.css` - minimized stylesheet with inline encoded images

## AngularJS

Web sources for AngularJS projects are stored in the `client` directory:

- `client/index.html` - Landing page source
- `client/component.html` - Landing page source
- `client/app.js` - Angular app initialization, config and routing
- `client/404.html` and `client/500.html` - Error page sources
- `client/default.css` - minimized stylesheet with inline encoded images



## Publishing Changes

In order to publish changes, you will need to fork the repository or branch off the `master` branch.

Make sure to follow the [conventional commit specification](https://conventionalcommits.org/) before contributing. To help you with commit a commit template is provide. Run `config.sh` to initialize the commit template to your `.git/config` or use [commitizen](https://www.npmjs.com/package/commitizen)

Once you are finished with your changes, run `npm test` to make sure all tests pass.

Do a pull request against `master`, make sure the build passes. A team member will review and merge your pull request.
Once merged to `master` an auto generated pull request will be created against master to update the changelog. Make sure that the CHANGELOG.md and the package.json is correct before merging the pull request. After the auto generated pull request has been merged to `master` the version will be bumped and published to npm.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.4.14"></a>
## [1.4.14](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.13...v1.4.14) (2020-02-26)



<a name="1.4.13"></a>
## [1.4.13](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.12...v1.4.13) (2019-10-09)



<a name="1.4.12"></a>
## [1.4.12](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.11...v1.4.12) (2019-09-03)



<a name="1.4.11"></a>
## [1.4.11](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.10...v1.4.11) (2019-05-17)



<a name="1.4.10"></a>
## [1.4.10](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.9...v1.4.10) (2019-02-14)


### Bug Fixes

* Add handlebars dependency ([5953a7a](https://github.com/ibm-developer/generator-ibm-web/commit/5953a7a))



<a name="1.4.9"></a>
## [1.4.9](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.8...v1.4.9) (2019-01-25)


### Bug Fixes

* backup uglify version ([0ad04f5](https://github.com/ibm-developer/generator-ibm-web/commit/0ad04f5))



<a name="1.4.8"></a>
## [1.4.8](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.7...v1.4.8) (2019-01-08)


### Bug Fixes

* rearrange a webpack devDep ([af8cd85](https://github.com/ibm-developer/generator-ibm-web/commit/af8cd85))



<a name="1.4.7"></a>
## [1.4.7](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.6...v1.4.7) (2019-01-03)


### Bug Fixes

* add webpack dependency ([d418581](https://github.com/ibm-developer/generator-ibm-web/commit/d418581))
* update cloud urls ([a149de3](https://github.com/ibm-developer/generator-ibm-web/commit/a149de3))



<a name="1.4.6"></a>
## [1.4.6](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.5...v1.4.6) (2018-12-19)


### Bug Fixes

* test-server.js only for Node ([18d387d](https://github.com/ibm-developer/generator-ibm-web/commit/18d387d))



<a name="1.4.5"></a>
## [1.4.5](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.4...v1.4.5) (2018-11-15)


### Bug Fixes

* Include lib dir in files so it shows up in npm ([f3561b6](https://github.com/ibm-developer/generator-ibm-web/commit/f3561b6)), closes [#59](https://github.com/ibm-developer/generator-ibm-web/issues/59)



<a name="1.4.4"></a>
## [1.4.4](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.3...v1.4.4) (2018-11-14)


### Bug Fixes

* added swift support to generator for landing pages ([0c19f69](https://github.com/ibm-developer/generator-ibm-web/commit/0c19f69))
* adding 404 and 500 pages to python microservice and blank starters ([9d201a2](https://github.com/ibm-developer/generator-ibm-web/commit/9d201a2))
* adding back index.html for node and python starters with backend focus ([92a3822](https://github.com/ibm-developer/generator-ibm-web/commit/92a3822))
* bff, microservice, and blank starters don't have index.html ([70cb8ac](https://github.com/ibm-developer/generator-ibm-web/commit/70cb8ac))
* fixed tests ([cd8ddf1](https://github.com/ibm-developer/generator-ibm-web/commit/cd8ddf1))
* refactored generator so all languages will draw from same template files for landing pages ([cdfaf1f](https://github.com/ibm-developer/generator-ibm-web/commit/cdfaf1f))



<a name="1.4.3"></a>
## [1.4.3](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.2...v1.4.3) (2018-10-26)


### Bug Fixes

* **test:** Use callback instead of timeout for server test ([763f37d](https://github.com/ibm-developer/generator-ibm-web/commit/763f37d))



<a name="1.4.2"></a>
## [1.4.2](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.1...v1.4.2) (2018-09-26)


### Bug Fixes

* update webpack-cli ([b7c17c0](https://github.com/ibm-developer/generator-ibm-web/commit/b7c17c0))



<a name="1.4.1"></a>
## [1.4.1](https://github.com/ibm-developer/generator-ibm-web/compare/v1.4.0...v1.4.1) (2018-09-14)


### Bug Fixes

* getting rid of ejs tags ([413c11c](https://github.com/ibm-developer/generator-ibm-web/commit/413c11c))
* update unit test for landing page ([3d5e97b](https://github.com/ibm-developer/generator-ibm-web/commit/3d5e97b))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/ibm-developer/generator-ibm-web/compare/v1.2.6...v1.4.0) (2018-08-20)


### Bug Fixes

* fix typo in package.json ([bc70a87](https://github.com/ibm-developer/generator-ibm-web/commit/bc70a87))
* package.json configuration ([0bd9bc0](https://github.com/ibm-developer/generator-ibm-web/commit/0bd9bc0))
* update unit test for landing page ([#39](https://github.com/ibm-developer/generator-ibm-web/issues/39)) ([ba2d889](https://github.com/ibm-developer/generator-ibm-web/commit/ba2d889))


### Features

* add golang functionality ([#37](https://github.com/ibm-developer/generator-ibm-web/issues/37)) ([d1cb247](https://github.com/ibm-developer/generator-ibm-web/commit/d1cb247))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/ibm-developer/generator-ibm-web/compare/v1.2.6...v1.3.0) (2018-08-01)


### Bug Fixes

* fix typo in package.json ([bc70a87](https://github.com/ibm-developer/generator-ibm-web/commit/bc70a87))
* package.json configuration ([0bd9bc0](https://github.com/ibm-developer/generator-ibm-web/commit/0bd9bc0))


### Features

* add golang functionality ([#37](https://github.com/ibm-developer/generator-ibm-web/issues/37)) ([d1cb247](https://github.com/ibm-developer/generator-ibm-web/commit/d1cb247))



<a name="1.2.7"></a>
## [1.2.7](https://github.com/ibm-developer/generator-ibm-web/compare/v1.2.6...v1.2.7) (2018-07-24)


### Bug Fixes

* fix typo in package.json ([bc70a87](https://github.com/ibm-developer/generator-ibm-web/commit/bc70a87))
* package.json configuration ([0bd9bc0](https://github.com/ibm-developer/generator-ibm-web/commit/0bd9bc0))



<a name="1.2.6"></a>
## [1.2.6](https://github.com/ibm-developer/generator-ibm-web/compare/v1.2.5...v1.2.6) (2018-06-22)

test/test.js
/*
 © Copyright IBM Corp. 2018
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

"use strict";

let helpers = require('yeoman-test');
let assert = require('yeoman-assert');
let path = require('path');
let fs = require('fs-extra');
const defaultNodeVersion = '^8.11.0';

let bluemixSettings = {
	name: "MyTest",
	backendPlatform: "SWIFT",
	"cloudant": [
		{
			"password": "pass",
			"url": "https://account.cloudant.com",
			"username": "user"
		}
	],
	"objectStorage": [
		{
			"password": "Gl.=W23@",
			"projectId": "12345",
			"region": "dallas",
			"userId": "abc1234"
		}
	]
};

let requiredFilesForBasic = [
	'public/index.html',
	'public/404.html',
	'public/500.html'
];

let requiredFilesForAngular = [
	'client/app.js',
	'client/component.html',
	'client/index.html',
	'public/404.html',
	'public/500.html',
	'client/default.css',
	'webpack.common.js',
	'webpack.dev-proxy.js',
	'webpack.dev-standalone.js',
	'webpack.prod.js',
];

let requiredFilesForReact = [
	'client/index.html',
	'client/app/App.jsx',
	'client/default.css',
	'client/index.jsx',
	'client/404.html',
	'client/500.html',
	'test/test-server.js',
	'Procfile-dev',
	'Procfile-debug',
	'webpack.common.js',
	'webpack.dev-proxy.js',
	'webpack.dev-standalone.js',
	'webpack.prod.js',
];

describe('Web project generator', function () {

	// NodeJS tests

	describe('Basic app with NodeJS', function () {

		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir()
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "None"
				});
		});

		it('contains web pages', function () {

			assert.file(requiredFilesForBasic);

		});

		it('starter text appears', function () {
			assert.fileContent('public/index.html', 'You are currently running a Node.js app built for the IBM Cloud')
		});

	});

	describe('Basic app with Swift', function () {

		beforeEach(function () {

			bluemixSettings.backendPlatform = "SWIFT";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir()
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "None"
				});
		});

		it('contains web pages', function () {

			assert.file(requiredFilesForBasic);

		});

		it('starter text appears', function () {
			assert.fileContent('public/index.html', 'You are currently running a Swift app built for the IBM Cloud')
		});

	});

	describe('Basic app with Python', function () {

		beforeEach(function () {

			bluemixSettings.backendPlatform = "PYTHON";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir()
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "None",
					appType: "WEB"
				});
		});

		it('contains web pages', function () {

			assert.file(requiredFilesForBasic);

		});

		it('starter text appears', function () {
			assert.fileContent('public/index.html', 'You are currently running a Python app built for the IBM Cloud')
		});

	});
	describe('Basic app with Go', function () {

		beforeEach(function () {

			bluemixSettings.backendPlatform = "GO";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir()
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "None"
				});
		});

		it('contains web pages', function () {

			assert.file(requiredFilesForBasic);

		});

		it('starter text appears', function () {
			assert.fileContent('public/index.html', 'You are currently running a Go app built for the IBM Cloud')
		});

	});

	describe('React app with NodeJS', function () {

		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir(function (dir) {
					fs.copySync(path.join(__dirname, 'resources/package.json'), path.join(dir, 'package.json'));
					fs.copySync(path.join(__dirname, 'resources/Dockerfile'), path.join(dir, 'Dockerfile'));
					fs.copySync(path.join(__dirname, 'resources/manifest.yml'), path.join(dir, 'manifest.yml'));
				})
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "React"
				});
		});

		it('required files created', function () {

			assert.file(requiredFilesForReact);

		});

		it('should have the correct run scripts to debug and dev', function() {
			assert.fileContent('package.json', '"dev": "nf --procfile Procfile-dev --port 3000 start"');
			assert.fileContent('package.json', '"debug": "nf --procfile Procfile-debug --port 3000 start"');
		});

		it('contains original dependencies', function () {

			assert.fileContent('package.json', 'appmetrics-dash');
			assert.fileContent('package.json', 'express');

		});

		it('contains Webpack', function () {

			assert.fileContent('package.json', 'webpack');
			assert.fileContent('package.json', 'babel');

		});

		it(`should use ${defaultNodeVersion} for engine node version`, function() {
			assert.jsonFileContent('package.json', {engines: {node : defaultNodeVersion}});
		});

		it('contains React', function () {
			assert.fileContent('package.json', 'react');
			assert.fileContent('package.json', 'react-dom');
			assert.fileContent('package.json', 'babel-preset-react');
		});

	});
	describe('React app with NodeJS', function () {

		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir(function (dir) {
					fs.copySync(path.join(__dirname, 'resources/package.json'), path.join(dir, 'package.json'));
					fs.copySync(path.join(__dirname, 'resources/Dockerfile'), path.join(dir, 'Dockerfile'));
					fs.copySync(path.join(__dirname, 'resources/manifest.yml'), path.join(dir, 'manifest.yml'));
				})
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "React"
				});
		});

		it('required files created', function () {

			assert.file(requiredFilesForReact);

		});

		it('should have the correct run scripts to debug and dev', function() {
			assert.fileContent('package.json', '"dev": "nf --procfile Procfile-dev --port 3000 start"');
			assert.fileContent('package.json', '"debug": "nf --procfile Procfile-debug --port 3000 start"');
		});

		it('contains original dependencies', function () {

			assert.fileContent('package.json', 'appmetrics-dash');
			assert.fileContent('package.json', 'express');

		});

		it('contains Webpack', function () {

			assert.fileContent('package.json', 'webpack');
			assert.fileContent('package.json', 'babel');

		});

		it(`should use ${defaultNodeVersion} for engine node version`, function() {
			assert.jsonFileContent('package.json', {engines: {node : defaultNodeVersion}});
		});

		it('contains React', function () {
			assert.fileContent('package.json', 'react');
			assert.fileContent('package.json', 'react-dom');
			assert.fileContent('package.json', 'babel-preset-react');
		});

	});
	describe('React app with NodeJS using a defined node version', function () {

		const expectedNodeVersion = '8.0.0';
		beforeEach(function () {
			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir(function (dir) {
					fs.copySync(path.join(__dirname, 'resources/Dockerfile'), path.join(dir, 'Dockerfile'));
					fs.copySync(path.join(__dirname, 'resources/manifest.yml'), path.join(dir, 'manifest.yml'));
				})
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "React",
					nodeVersion: expectedNodeVersion
				});
		});

		it('required files created', function () {

			assert.file(requiredFilesForReact);

		});

		it('contains Webpack', function () {

			assert.fileContent('package.json', 'webpack');
			assert.fileContent('package.json', 'babel');

		});

		it('contains React', function () {
			assert.fileContent('package.json', 'react');
			assert.fileContent('package.json', 'react-dom');
			assert.fileContent('package.json', 'babel-preset-react');
		});
	});

	describe('AngularJS app with NodeJS using a defined node version and present existing files ', function () {
		const expectedNodeVersion = '8.0.0';
		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir(function (dir) {
					fs.copySync(path.join(__dirname, 'resources/Dockerfile'), path.join(dir, 'Dockerfile'));
					fs.copySync(path.join(__dirname, 'resources/manifest.yml'), path.join(dir, 'manifest.yml'));
				})
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "AngularJS",
					nodeVersion: expectedNodeVersion
				});
		});


		it('required files created', function () {
			assert.file(requiredFilesForAngular);
		});

		it('should have specific build script', function() {
			assert.fileContent('package.json', 'webpack --progress --config webpack.prod.js');
		});

		it('should have correct node scripts under dev and debug', function() {
			assert.fileContent('package.json', '"dev": "npm-run-all --parallel client-reload-proxy server-reload"');
			assert.fileContent('package.json', '"debug": "npm-run-all --parallel client-reload-proxy inspector"');
		});

		it('should have original scripts and dependencies', function() {
			assert.fileContent('package.json', 'mocha');
			assert.fileContent('package.json', 'node --inspect=0.0.0.0:9229 server/server.js');
		});
	});
	describe('AngularJS app with NodeJS and present existing files ', function () {
		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.inTmpDir(function (dir) {
					fs.copySync(path.join(__dirname, 'resources/package.json'), path.join(dir, 'package.json'));
					fs.copySync(path.join(__dirname, 'resources/Dockerfile'), path.join(dir, 'Dockerfile'));
					fs.copySync(path.join(__dirname, 'resources/manifest.yml'), path.join(dir, 'manifest.yml'));
				})
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "AngularJS",
				});
		});

		it('required files created', function () {
			assert.file(requiredFilesForAngular);
		});

		it('contains original dependencies', function () {

			assert.fileContent('package.json', 'appmetrics-dash');
			assert.fileContent('package.json', 'express');

		});

		it('should have react specific build script', function() {
			assert.fileContent('package.json', 'webpack --progress --config webpack.prod.js');
		})

		it('should have correct node scripts under dev and debug', function() {
			assert.fileContent('package.json', '"dev": "npm-run-all --parallel client-reload-proxy server-reload"');
			assert.fileContent('package.json', '"debug": "npm-run-all --parallel client-reload-proxy inspector"');
		});

		it(`should use ${defaultNodeVersion} for engine node version`, function() {
			assert.jsonFileContent('package.json', {engines: {node : defaultNodeVersion}});
		});
		it('should have original scripts and dependencies', function() {
			assert.fileContent('package.json', 'mocha');
			assert.fileContent('package.json', 'node --inspect=0.0.0.0:9229 server/server.js');
			assert.fileContent('package.json', 'sample-app');
			assert.jsonFileContent('package.json', {engines: {node : defaultNodeVersion}});
		});
	});

	describe('AngularJS app with NodeJS with empty user directory', function () {
		beforeEach(function () {

			bluemixSettings.backendPlatform = "NODE";

			return helpers.run(path.join(__dirname, '../generators/app'))
				.withOptions({
					bluemix: JSON.stringify(bluemixSettings),
					framework: "AngularJS"
				});
		});

		it('required files created', function () {
			assert.file(requiredFilesForAngular);
		});

		it('contains original dependencies', function () {

			assert.fileContent('package.json', 'appmetrics-dash');
			assert.fileContent('package.json', 'express');

		});

		it(`should use ${defaultNodeVersion} for engine node version`, function() {
			assert.fileContent('package.json', 'MyTest');
			assert.jsonFileContent('package.json', {engines: {node : defaultNodeVersion}});
		});

		it('should not have Dockerfile', function () {
			assert.noFile('Dockerfile')
		});

		it('should not have manifest.yml', function () {
			assert.noFile('manifest.yml');
		});

		it('should have react specific build script', function() {
			assert.fileContent('package.json', 'webpack --progress --config webpack.prod.js');
		});

		it('should have correct node scripts under dev and debug', function() {
			assert.fileContent('package.json', '"dev": "npm-run-all --parallel client-reload-proxy server-reload"');
			assert.fileContent('package.json', '"debug": "npm-run-all --parallel client-reload-proxy inspector"');
		});

		it('should have original scripts and dependencies', function() {
			assert.fileContent('package.json', 'mocha');
			assert.fileContent('package.json', 'node --inspect=0.0.0.0:9229 server/server.js');
		})
	});
});


lib/handlebars.js
/*
 © Copyright IBM Corp. 2018
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */


'use strict';

const Handlebars = require('handlebars');

Handlebars.registerHelper('ifCond', function (v1, operator, v2, options) {

	switch (operator) {
		case '===':
			return (v1 === v2) ? options.fn(this) : options.inverse(this);
		default:
			return options.inverse(this);
	}
});

module.exports = Handlebars;


generators/app/index.js
/*
 © Copyright IBM Corp. 2018
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
 http://www.apache.org/licenses/LICENSE-2.0
 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

'use strict';
const Generator = require('yeoman-generator');
const dep = require('./templates/dependencies.json');
const reactDep = require('./templates/react/dependencies.json');
const angularJsDep = require('./templates/angularjs/dependencies.json');
const Handlebars = require('../../lib/handlebars.js');

const scripts = dep.scripts;
const angularJsScripts = angularJsDep.scripts;

const devDependencies = dep.devDependencies;
const angularJsDevDependencies = angularJsDep.devDependencies;
const defaultNodeVersion = "^8.11.0";

module.exports = class extends Generator {
	constructor(args, opts) {
		super(args, opts);

		this.framework = opts.framework;
		this.nodeVersion = opts.nodeVersion || defaultNodeVersion;

		if (typeof (opts.bluemix) === 'string') {
			this.bluemix = JSON.parse(opts.bluemix || "{}");
		} else {
			this.bluemix = opts.bluemix;
		}

		if (opts.cloudContext) {
			this.opts = opts.cloudContext
			this.opts.libertyVersion = opts.libertyVersion
		} else {
			this.opts = opts
		}

		this.humanNameLanguage = {
			"NODE": "Node.js",
			"SWIFT": "Swift",
			"JAVA": "Java",
			"PYTHON": "Python",
			"DJANGO": "Django",
			"GO": "Go"
		};
	}

	paths() {
		this.sourceRoot();
		this.templatePath('index.js');
	}

	prompting() {
		let prompts = [];
		if (this.bluemix === undefined) {
			this.bluemix = {};
			prompts.push({
				type: 'input',
				name: 'name',
				message: 'Your project name'
			});
			prompts.push({
				type: 'list',
				name: 'framework',
				message: 'What framework are you using',
				choices: [
					"None",
					"React",
					"AngularJS"
				]
			});
			prompts.push({
				type: 'list',
				name: 'language',
				message: 'What language are you using',
				choices: [
					"SWIFT",
					"NODE",
					"JAVA",
					"PYTHON",
					"DJANGO",
					"GO"
				]
			});

			prompts.push({
				type: 'input',
				name: 'nodeVersion',
				when: answers => answers.language === 'NODE',
				message: 'Specify the verison of node that you wish to use.'
			});

		}

		return this.prompt(prompts).then(this._processAnswers.bind(this));
	}

	_processAnswers(answers) {

		this.bluemix.backendPlatform = answers.language || this.bluemix.backendPlatform;
		this.framework = answers.framework || this.framework;
		this.bluemix.name = answers.name || this.bluemix.name;
		this.nodeVersion = answers.nodeVersion || this.nodeVersion;
	}

	write() {
		switch (this.framework) {
			case "None":
				this._generateBasic();
				break;
			case "React":
				this._generateReact();
				break;
			case "AngularJS":
				this._generateAngularJS();
				break;
			default:
				this._generateBasic();
				break;
		}
	}

	_generateBasic() {

		// Replace server test with Web specific test.
		if (this.bluemix.backendPlatform === 'NODE') {
			this.fs.copyTpl(
				this.templatePath('basic/test-server.js'),
				this.destinationPath('test/test-server.js'), {}
			);
		}
		this._writeHandlebarsFile('basic/index.html', 'public/index.html', {
			applicationName: this.bluemix.name,
			language: this.humanNameLanguage[this.bluemix.backendPlatform]
		});
		this._writeHandlebarsFile('basic/404.html', 'public/404.html', {
			applicationName: this.bluemix.name,
			language: this.humanNameLanguage[this.bluemix.backendPlatform]
		});
		this._writeHandlebarsFile('basic/500.html', 'public/500.html', {
			applicationName: this.bluemix.name,
			language: this.humanNameLanguage[this.bluemix.backendPlatform]
		});

	}

	_generateReact() {

		// Replace server test with Web specific test.
		this.fs.copyTpl(
			this.templatePath('react/test-server.js'),
			this.destinationPath('test/test-server.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/Procfile-dev'),
			this.destinationPath('Procfile-dev'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/Procfile-debug'),
			this.destinationPath('Procfile-debug'), {}
		);


		this.fs.copyTpl(
			this.templatePath('react/webpack.common.js'),
			this.destinationPath('webpack.common.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/webpack.dev-proxy.js'),
			this.destinationPath('webpack.dev-proxy.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/webpack.dev-standalone.js'),
			this.destinationPath('webpack.dev-standalone.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('react/webpack.prod.js'),
			this.destinationPath('webpack.prod.js'), {}
		);

		if (!this.fs.exists(this.destinationPath('package.json'))) {
			this._writeHandlebarsFile('package.json', 'package.json', {
				applicationName: this.bluemix.name,
				language: this.bluemix.backendPlatform,
				nodeVersion: this.nodeVersion
			});
		}

		this._augmentPackageJSON({ react: true });

		this.fs.copyTpl(
			this.templatePath('react/client'),
			this.destinationPath('client'), {}
		);

	}

	_generateAngularJS() {

		// Replace server test with Web specific test.
		this.fs.copyTpl(
			this.templatePath('angularjs/test-server.js'),
			this.destinationPath('test/test-server.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/webpack.common.js'),
			this.destinationPath('webpack.common.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/webpack.dev-proxy.js'),
			this.destinationPath('webpack.dev-proxy.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/webpack.dev-standalone.js'),
			this.destinationPath('webpack.dev-standalone.js'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/webpack.prod.js'),
			this.destinationPath('webpack.prod.js'), {}
		);

		if (!this.fs.exists(this.destinationPath('package.json'))) {
			this._writeHandlebarsFile('package.json', 'package.json', {
				applicationName: this.bluemix.name,
				language: this.bluemix.backendPlatform,
				nodeVersion: this.nodeVersion
			});
		}

		this._augmentPackageJSON({ angularjs: true });

		this.fs.copyTpl(
			this.templatePath('angularjs/client/app.js'),
			this.destinationPath('client/app.js'), {}
		);
		this.fs.copyTpl(
			this.templatePath('angularjs/client/component.html'),
			this.destinationPath('client/component.html'), {}
		);
		this.fs.copyTpl(
			this.templatePath('angularjs/client/default.css'),
			this.destinationPath('client/default.css'), {}
		);
		this.fs.copyTpl(
			this.templatePath('angularjs/client/index.html'),
			this.destinationPath('client/index.html'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/client/404.html'),
			this.destinationPath('public/404.html'), {}
		);

		this.fs.copyTpl(
			this.templatePath('angularjs/client/500.html'),
			this.destinationPath('public/500.html'), {}
		);

	}

	_augmentPackageJSON(options) {
		let packageFile = this.fs.read(this.destinationPath('package.json'), { defaults: "{}" });
		let packageFileJSON;

		try {
			packageFileJSON = JSON.parse(packageFile);
		} catch (err) {
			packageFileJSON = {};
		}

		if (packageFileJSON.devDependencies === undefined) {
			packageFileJSON.devDependencies = {};
		}

		packageFileJSON.scripts = Object.assign(packageFileJSON.scripts, scripts);
		packageFileJSON.devDependencies = Object.assign(packageFileJSON.devDependencies, devDependencies);

		if (options !== undefined && options.react) {
			packageFileJSON.scripts = Object.assign(packageFileJSON.scripts, reactDep.scripts);
			packageFileJSON.devDependencies = Object.assign(packageFileJSON.devDependencies, reactDep.devDependencies);
		}

		if (options !== undefined && options.angularjs) {
			packageFileJSON.devDependencies = Object.assign(packageFileJSON.devDependencies, angularJsDevDependencies);
			packageFileJSON.scripts = Object.assign(packageFileJSON.scripts, angularJsScripts);
		}

		this.fs.writeJSON(this.destinationPath('package.json'), packageFileJSON, null, 4);
	}

	_writeHandlebarsFile(templateFile, destinationFile, data) {
		let template = this.fs.read(this.templatePath(templateFile));
		let compiledTemplate = Handlebars.compile(template);
		let output = compiledTemplate(data);
		this.fs.write(this.destinationPath(destinationFile), output);
	}


};
